nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/AttributesToCSV.java

```diff
@@ -0,0 +1,280 @@
+        return relationships;
+    }
+
+
+    private Map<String, String> buildAttributesMapForFlowFile(FlowFile ff, Set<String> userSpecifiedAttributes) {
+        Map<String, String> result;
+        if (!userSpecifiedAttributes.isEmpty()) {
+            //the user gave a list of attributes
+            result = new HashMap<>(userSpecifiedAttributes.size());
+            for (String attribute : userSpecifiedAttributes) {
+                String val = ff.getAttribute(attribute);
+                if (val != null && !val.isEmpty()) {
+                    result.put(attribute, val);
+                } else {
+                    if (nullValForEmptyString) {
+                        result.put(attribute, "null");
+                    } else {
+                        result.put(attribute, "");
+                    }
+                }
+            }
+        } else {
+            //the user did not give a list of attributes, take all the attributes from the flowfile
+            Map<String, String> ffAttributes = ff.getAttributes();
+            result = new HashMap<>(ffAttributes.size());
+            for (Map.Entry<String, String> e : ffAttributes.entrySet()) {
+                    result.put(e.getKey(), e.getValue());
+            }
+        }
+
+        //now glue on the core attributes if the user wants them.
+        if(includeCoreAttributes) {
+            for (String coreAttribute : coreAttributes) {
+                String val = ff.getAttribute(coreAttribute);
+                //make sure this coreAttribute is applicable to this flowfile.
+                if(ff.getAttributes().containsKey(coreAttribute)) {
+                    if (val != null && !val.isEmpty()) {
+                        result.put(coreAttribute, val);
+                    } else {
+                        if (nullValForEmptyString) {
+                            result.put(coreAttribute, "null");
+                        } else {
+                            result.put(coreAttribute, "");
+                        }
+                    }
+                }
+            }
+        } else {
+            //remove core attributes since the user does not want them, unless they are in the attribute list. Attribute List always wins
+            for (String coreAttribute : coreAttributes) {
+                //never override user specified attributes, even if the user has selected to exclude core attributes
+                if(!userSpecifiedAttributes.contains(coreAttribute)) {
+                    result.remove(coreAttribute);
+                }
+            }
+        }
+        return result;
+    }
+
+    private Set<String> attributeListStringToSet(String attributeList) {
+        //take the user specified attribute list string and convert to list of strings.
+        Set<String> result = new HashSet<>();
+        if (StringUtils.isNotBlank(attributeList)) {
+            String[] ats = StringUtils.split(attributeList, OUTPUT_SEPARATOR);
```

Comment:
Another edge case (sorry), attribute names are allowed to have commas in them :( I'm totally fine with not supporting that kind of thing, but maybe add to the description of Attribute List that attribute names containing commas are not allowed. Or if you want to support them, then Attribute List would itself be in CSV format and would have to be parsed as such.  I'm good either way, just pointing it out for completeness.